### ISSUE TITLE
Reading GUID of external file

### ISSUE THREAD

**Issue by Eneroth3** [SketchUp/api-issue-tracker#225]
In SketchUp GUIDs are used to identify identical component definitions. For each change to a component a new GUID is assigned. If you copy a component into a file and there is already a component definition with the same GUID, it is re-used. If no component definitioon with that GUID exists, the copied component gets added to the model, and has it's name updated if it collides with an existing component. For instance if you create a component named "Window", copy it to a new file, edit it and copy it back it gets renamed "Window#1" and the old component is kept as it is. If you never edited the component you'll only have "Window#1".

However, it doesn't seem possible to read the GUID of an external file through the API. This would be highly useful for component referencing, to know if the external model is identical to the internal component.

Potentially this can already be hacked out of the file using a similar method as for reading file version (see #30) but I haven't had the time to check it yet.

**Comment by DanRathbun** [comment]
Probably the safest means is to use the C API and have a C Ruby extension.

In the long run, it'd be nice to have a special **`Sketchup::SKP`** class for handling external files similar to the standard Ruby library's [`CSV`](http://ruby-doc.org/stdlib-2.5.1/libdoc/csv/rdoc/index.html) class. Perhaps this could be an open source community project ? (It'd serve as a nifty C extension example as well.)

**Comment by Eneroth3** [comment]
I agree using API is safest here. For the SketchUp file version the data needs to be easily accessible in a backward compatible way, which can also be read directly in Ruby. The GUID on the other hand may very well be berried deep and require the whole file to be parsed according to a version specific standard to be dug up.

Currently there is no function in the C SDK to get the GUID of a model but it seems I can create a through away model, insert the model I am interesting in as a component and read the GUID for the component definition.

Having a class in the Ruby API for getting GUIDs, file version and other data from external files would sure be useful! No need to recompile the extension for each new SU version.

**Comment by Eneroth3** [comment]
I've headached by way through the C API documentation but can't find a way to load a component into a model. Is this currently possible? Comments to other issues suggests not.

**Comment by DanRathbun** [comment]
I also do not see a "load" function. But we've noticed this before it seems, _see:_
* [Component loading and reloading #50](https://github.com/SketchUp/api-issue-tracker/issues/50)

There is the [`SUModelCreateFromFile`](http://extensions.sketchup.com/developer_center/sketchup_c_api/sketchup/model_8h.html#a14af4587c84761d26dfdee908e73d9cd) function.

To mimic a definition load, we'd have to open both files, and recreate the component file's entities into a new definition in the target file. Tedious for sure. And this would cause a new GUID, I suppose.

> Currently there is no function in the C SDK to get the GUID of a model ...

Okay I see it also has been neglected and requested previously ...

* [CAPI Feature request: SUModelGetGuid #58](https://github.com/SketchUp/api-issue-tracker/issues/58)

> The GUID on the other hand may very well be berried deep and require the whole file to be parsed according to a version specific standard to be dug up.

I do not see any plain text GUID in the file that can be easily "hacked" out by parsing the file using Ruby. I'd prefer a dedicated C API function.

**Comment by Eneroth3** [comment]
It might be worth looking into if this data can be hacked out of the skp file in the meantime, even if it isn't an ideal solution.

I have vague memories of having seen skp files open as archive, either in Gmail/Google Docs or when I have the skp file inside another archive and double click it in a program like 7zip. I've tried this again but haven't got it to work. 7zip just tells me the skp file isn't an archive when i try to open it as one.

Maybe I'm mixing up skp files with layout files.

In any case it wouldn't be a long term solution, just a temporary hack to get the job done for now.

**Comment by DanRathbun** [comment]
SKP is a binary format, and as I noted above (having opened up the file as plain text) ...

> I do not see any plain text GUID in the file that can be easily "hacked" out by parsing the file using Ruby.

This means the GUID is not stored in plain ASCII text. It might be unicode, or it might be encoded in some way. I did not actually look at a SKP file's binary format. So anyway I have not yet found how to "hack" it from the file data.

**Comment by thomthom** [comment]
> However, it doesn't seem possible to read the GUID of an external file through the API.

I read this thread as a request to obtain the model GUID in both Ruby and C API without having to read/load the entire file. (?)

Meanwhile I can shed some light on how the GUID is stored in the SKP file; (Beware - this could be subject to change - the SKP file format is in no form part of the API contract.)

It's tricky to read .skp files manually. In large parts data is contained within blocks of data prefixed with `FF FE FF` followed by another byte indicating the size of the data block. The name of the datablock is in UTF-16 format (`wchar_t`).

The very first data block is the file type id, simply a block named "SketchUp Model".
Then there's a version info block.
Directly after the version info are 16 bytes that represent the GUID. (Note that this is not in the block data signature - it's just 16bytes dumped directly after the version info.)

> The GUID on the other hand may very well be berried deep and require the whole file to be parsed according to a version specific standard to be dug up.

There is a code comment next to where we serialize the GUID:
```
    // Save the guid
    // Do this near the beginning of the file so we can get it
    // from a file without reading the whole thing.
```

So it's intentionally put in the beginning of the file. :)

The serialized GUID is at hart a struct defined as:

```
struct Guid {
  unsigned int data1;
  unsigned short data2;
  unsigned short data3;
  unsigned char data4[8];
}
```

The string representation you get from `model.guid` is a `sprintf` formatting of the values into hexadecimal with some dashes sprinkled into the mix.

To read this in Ruby you can do something like this:

```ruby
def read_model_guid(path)
  guid = File.open(path, 'rb') do |file|
    # SKP File Header Structure:
    #
    # "SketchUp Model" ID block (Fixed 32 bytes)
    # File version block (Variable bytes)
    # GUID (Fixed 16 bytes)
    #
    # Each block starts with FF FE FF followed by the block name size.

    # First skip over the .skp ID block and fetch the size of the version info.
    file.seek(32 + 3)
    version_wchar_size = file.read(1).unpack('C').first
    version_byte_size = version_wchar_size * 2

    # Skip the version info.
    file.seek(version_byte_size, IO::SEEK_CUR)

    # Read the raw GUID bytes.
    #
    # struct Guid {
    #   unsigned int data1;
    #   unsigned short data2;
    #   unsigned short data3;
    #   unsigned char data4[8];
    # }
    file.read(16)
  end
  # Unpack the GUID into a human readable string.
  data1, data2, data3, *data4 = guid.unpack('ISSC8')
  sprintf("%.8x-%.4x-%.4x-%.2x%.2x-%.2x%.2x%.2x%.2x%.2x%.2x",
     data1, data2, data3, *data4)
end
```

Quick little snippet to verify in the Ruby Console of the file you have open:

```ruby
model = Sketchup.active_model
guid = read_model_guid(model.path)

puts
p model.guid
p guid
p guid == model.guid
puts
```

**Comment by Eneroth3** [comment]
This is wonderful! Even if it isn't official API I can live with it; I'll just through in some unit tests to easily know if it changes in a future SU version.

When thinking of it this probably isn't very likely to change either, given how SketchUp is able to read it without parsing the whole file.

**Comment by DanRathbun** [comment]
> I read this thread as a request to obtain the model GUID in both Ruby and C API without having to read/load the entire file. (?)

Yes. true, external file. We already have `Sketchup::Model#guid` in the Ruby API to get the active loaded model's guid, and the complimentary method for component/group definition's guid.

However as noted above and in issue #58, the **C API** has no current function or means to read the model object's guid. (I could not see a way to type cast the model object to a definition in order to use it's [`SUComponentDefinitionGetGuid`](http://extensions.sketchup.com/developer_center/sketchup_c_api/sketchup/component__definition_8h.html#a68a81c30afb9026f724f2cae48cf8245) function.)

----

>To read this in Ruby you can do something like this ...

**Fantastic! Amazing what you can do if you know the internals.**

I find it interesting that it didn't throw an **`IOError`** on the SKP file that was already open.

Also if the active model has not yet been saved (the path string is empty) then the snippet will fail ...
```text
Error: #<Errno::ENOENT: No such file or directory @ rb_sysopen - >
<main>:in `initialize'
<main>:in `open'
```

**Comment by thomthom** [comment]
> Fantastic! Amazing what you can do if you know the internals.

I had many years ago reversed engineered parts of the SKP format. I wanted to extract information in a PHP server. Curiously, the GUID part of the header was one of the things I had not worked out what it was. I saw I changing on each save - I was wondering if it was some kind of timestamp or something. Digging into this now answered one of my own old questions. :D

> I find it interesting that it didn't throw an `IOError` on the SKP file that was already open.

You'd probably get an error if you opened the file for writing. OSs allow you to open files in various ways - among them is to open a file and lock it for editing but still allow other to still read it.

**Comment by thomthom** [comment]
> This is wonderful! Even if it isn't official API I can live with it;

I see the use for having ways to obtain info from SKP files without reading the whole file. Among them are GUID and thumbnail. Might be some other meta information such as title and description that could be useful.

**Comment by Eneroth3** [comment]
Glue to, always face camera and shadow face sun might be useful too. If you make a custom component finder you may want to filter by such properties. In Model Info all these properties are placed at the File page. Geo-location could be useful too but if I'm not mistaken it's saved as attributes on the Model object and would probably require the file to be parsed.

**Comment by thomthom** [comment]
> Glue to, always face camera and shadow face sun
That's part of the component/model behaviour - also something that is early in the file format.

**Comment by Eneroth3** [comment]
I've added this method, along with reading file version and checking if file is a SketchUp model, to the [SketchUp extension library](https://github.com/Eneroth3/sketchup-community-lib/blob/master/modules/skp.rb) I'm maintaining.

There are unit tests too for all these methods.

**Comment by Eneroth3** [comment]
Just an example what knowing the GUID of external files can be useful for:

![2019-04-08_12h25_29](https://user-images.githubusercontent.com/5631725/55717366-7bc8d980-59f9-11e9-903a-0e96d175fd44.gif)

**Comment by thomthom** [comment]
Tracked as SU-47318

**Comment by DanRathbun** [comment]
@thomthom Need "Fix-2021.0" label.
